test/OpenTelemetry.Api.Tests/BaggageTests.cs

```diff
@@ -266,10 +274,8 @@ public async Task AsyncLocalTests()
 
         Baggage.SetBaggage("key4", "value4");
 
-        Assert.Equal(4, Baggage.Current.Count);
+        Assert.Equal(2, Baggage.Current.Count);
         Assert.Equal("value1", Baggage.GetBaggage("key1"));
-        Assert.Equal("value2", Baggage.GetBaggage("key2"));
-        Assert.Equal("value3", Baggage.GetBaggage("key3"));
```

Comment:
To some, this could seem like a very nuanced and perhaps unintuitive behavioral difference that stems from a call to an async method versus a synchronous method call. For documentation purposes, I suggest adding a synchronous test that looks almost identical just to further highlight the nuanced behavior. Something like:

```csharp
    [Fact]
    public void NonAsyncLocalTests()
    {
        Baggage.SetBaggage("key1", "value1");

        InnerTask();

        Baggage.SetBaggage("key4", "value4");

        Assert.Equal(4, Baggage.Current.Count);
        Assert.Equal("value1", Baggage.GetBaggage("key1"));
        Assert.Equal("value2", Baggage.GetBaggage("key2"));
        Assert.Equal("value3", Baggage.GetBaggage("key3"));
        Assert.Equal("value4", Baggage.GetBaggage("key4"));

        static void InnerTask()
        {
            Baggage.SetBaggage("key2", "value2");
            Baggage.SetBaggage("key3", "value3");

            Assert.Equal(3, Baggage.Current.Count);
            Assert.Equal("value1", Baggage.GetBaggage("key1"));
            Assert.Equal("value2", Baggage.GetBaggage("key2"));
            Assert.Equal("value3", Baggage.GetBaggage("key3"));
        }
    }
```

That said, though... while I understand the nuanced behavior, how would we want to explain this difference to end users? Can you add some additional description of the why in the PR description?

Reply:
Sharing my speculation on what I think (most?) users would intuitively expect in few different scenarios. Admittedly these things can be very subjective, however, this is based on my own experience supporting product features similar to the baggage api.

```
    public void BaggageBehaviorWithSyncCall()
    {
        Baggage.SetBaggage("key1", "value1");
        InnerTask();
        Assert.Equal(2, Baggage.Current.Count);
        Assert.Equal("value1", Baggage.GetBaggage("key1"));
        Assert.Equal("value2", Baggage.GetBaggage("key2"));

        static void InnerTask()
        {
            Baggage.SetBaggage("key2", "value2");
        }
    }

    public async Task BaggageBehaviorWithAsyncCall()
    {
        Baggage.SetBaggage("key1", "value1");
        await InnerTask();
        Assert.Equal(2, Baggage.Current.Count);
        Assert.Equal("value1", Baggage.GetBaggage("key1"));
        Assert.Equal("value2", Baggage.GetBaggage("key2"));

        static async Task InnerTask()
        {
            Baggage.SetBaggage("key2", "value2");
            await Task.Yield();
        }
    }

    public async Task BaggageBehaviorWithTaskRun()
    {
        Baggage.SetBaggage("key1", "value1");
        Task.Run(InnerTask).GetAwaiter().GetResult();
        Assert.Equal(1, Baggage.Current.Count);
        Assert.Equal("value1", Baggage.GetBaggage("key1"));

        static void InnerTask()
        {
            Baggage.SetBaggage("key2", "value2");
        }
    }
```

I think my `Task.Run` scenario represents what folks are after in #3257 and I agree that it is the ideal behavior in this specific scenario most of the time - i.e., a parallel operation should ideally have it's own context. I'd go as far to say that this expectation is sometimes not limited to just baggage, but also to trace context, etc.

When it comes to `async/await` calls, we're often not talking about a parallel operation in the CPU-bound kind of sense, but rather an I/O-bound operation that is logically considered part of the same unit of work as the caller. In this way, I find that people expect this scenario to behave in the same way as though they had made a synchronous call.

Reply:
> When it comes to `async/await` calls, we're often not talking about a parallel operation in the CPU-bound kind of sense, but rather an I/O-bound operation that is logically considered part of the same unit of work as the caller. In this way, I find that people expect this scenario to behave in the same way as though they had made a synchronous call.

This is also subjective and murky - for example, an async/await call to an authentication library which interacts with backend services. Here is my suggestion when we encounter API design challenges that are subjective:
1. Follow the specification, if the spec does not make sense, propose a fix to the spec (which is good forcing function to make people think deeper).
2. Prioritize **fail-fast**. In this case, if the async/await call set the Baggage inside a Task and the developers were expecting the Baggage update from the caller's view, they will immediately notice things didn't work then figure out why. On the other side, context/baggage leakage is very hard to troubleshoot and would normally be discovered at much later stage of the development.
3. Provide explicit way for people to achieve special scenarios, e.g. here are some potential solutions https://github.com/open-telemetry/opentelemetry-dotnet/issues/3257#issuecomment-1883606017

Reply:
> Follow the specification, if the spec does not make sense, propose a fix to the spec (which is good forcing function to make people think deeper).

Fully agree with all of these principles, but I think this is the most pertinent one at the moment. I am not yet convinced that this PR brings us in closer alignment with the spec, but I do agree that it may be because the spec needs to be fixed if folks think this PR is a more correct behavior. As it stands, I do not find the new behavior proposed in this PR is any less murky or subjective.

Related, there exists critique that OpenTelemetry .NET does not implement the [Context spec](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/context) at all. See: https://github.com/open-telemetry/oteps/pull/232#discussion_r1228133487.

I think we should hold on this PR and do some deeper thinking.

> Provide explicit way for people to achieve special scenarios, e.g. here are some potential solutions Baggage.Current gets shared/overwritten when running multiple parallel tasks? #3257 (comment)

I support @CodeBlanch's workaround here https://github.com/open-telemetry/opentelemetry-dotnet/issues/3257#issuecomment-1882061279

> there is a workaround you can use right now: Call RuntimeContext.SetValue("otel.baggage", null)

This is effectively how `Activity.Current` is designed. If I spin up parallel work and wish to force a new trace I can set `Activity.Current` to null at the start of my new thread. To me, it makes more sense to try and keep the behavior of `Baggage.Current` aligned with the behavior of `Activity.Current`.

Reply:
We _could_ be more creative. Here is an idea: https://github.com/CodeBlanch/opentelemetry-dotnet/compare/api-baggage-context...CodeBlanch:opentelemetry-dotnet:api-baggage-context-flow?expand=1

What I'm doing there is add an enum so users can explicitly control the behavior.

```csharp
public async Task InnerTask()
{
    // Explicitly fork baggage for downstream call
    Baggage.SetBaggage("key2", "value2", flowBehavior: BaggageFlowBehavior.Fork);

    await CallSomething();
}

public async Task InnerTask()
{
    // Modify baggage as it works today. This will flow back
    Baggage.SetBaggage("key2", "value2", flowBehavior: BaggageFlowBehavior.ModifyCurrent);

    await CallSomething();
}

public async Task InnerTask()
{
    // Whatever the default behavior is. Right now it is Fork
    Baggage.SetBaggage("key2", "value2");

    await CallSomething();
}
```

Reply:
Here is another creative solution: https://github.com/CodeBlanch/opentelemetry-dotnet/compare/api-baggage-context...CodeBlanch:opentelemetry-dotnet:api-baggage-context-mutate?expand=1

This is sort of an attempt to do what Reiley proposed ("Get a handle of the current baggage and pass the handle to other methods"):

```csharp
        Baggage.SetBaggage("appInstanceId", "abcd");

        await ReadTenantIdFromDatabase();

        // Baggage.Current here will contain "appInstanceId" & "tenantId"

        async Task ReadTenantIdFromDatabase()
        {
            // Simulate async call to DB
            await Task.Yield();

            Baggage.MutateCurrentBaggage(b => b.SetBaggage("tenantId", "1234"));
        }
```